src/imagecompositor.cpp

```diff
@@ -200,95 +211,80 @@ void ImageCompositor::calibrate_linear(QImage &image, double a, double b) {
 }
 
 void ImageCompositor::getChannel(QImage &image, size_t channel) {
-    if (image.format() != QImage::Format_Grayscale16) {
-        image = QImage(image.width(), image.height(), QImage::Format_Grayscale16);
-    }
-    std::memcpy(image.bits(), rawChannels[channel-1].bits(), sizeof(unsigned short) * m_width * m_height);
-
-    if (m_isFlipped) {
-        image = image.mirrored(true, true);
-    }
+    image = rawChannels[channel-1];
 }
 
 // Composites 3 Grayscale16 images into each channel of a RGB16 image
 void ImageCompositor::getComposite(QImage &image, std::array<size_t, 3> chs) {
     if (image.format() != QImage::Format_RGBX64) {
         image = QImage(image.width(), image.height(), QImage::Format_RGBX64);
     }
-    QRgba64 *bits = reinterpret_cast<QRgba64 *>(image.bits());
 
-    quint16 *chbits[3];
-    for (size_t i = 0; i < 3; i++) {
-        chbits[i] = reinterpret_cast<quint16 *>(rawChannels[chs[i]-1].bits());
-    }
+    for (size_t i = 0; i < m_height; i++) {
+        QRgba64 *line = (QRgba64 *)image.scanLine(i);
+        const uint16_t *r = (const uint16_t *)rawChannels[chs[0]-1].constScanLine(i);
+        const uint16_t *g = (const uint16_t *)rawChannels[chs[1]-1].constScanLine(i);
+        const uint16_t *b = (const uint16_t *)rawChannels[chs[2]-1].constScanLine(i);
 
-    #pragma omp parallel for
-    for (size_t i = 0; i < m_height*m_width; i++) {
-        bits[i] = QRgba64::fromRgba64(chbits[0][i], chbits[1][i], chbits[2][i], UINT16_MAX);
-    }
-
-    if (m_isFlipped) {
-        image = image.mirrored(true, true);
+        #pragma omp parallel for
+        for (size_t x = 0; x < m_width; x++) {
+            line[x] = QRgba64::fromRgba64(r[x], g[x], b[x], UINT16_MAX);
+        }
     }
 }
 
 // Evaluate `expression` and store the results in `image`
-void ImageCompositor::getExpression(QImage &image, std::string experssion) {
-    QImage::Format format = (experssion.find(",") == std::string::npos) ? QImage::Format_Grayscale16 : QImage::Format_RGBX64;
+void ImageCompositor::getExpression(QImage &image, std::string expression) {
+    QImage::Format format = (expression.find(",") == std::string::npos) ? QImage::Format_Grayscale16 : QImage::Format_RGBX64;
     if (image.format() != format) {
         image = QImage(image.width(), image.height(), format);
     }
 
-    std::vector<quint16 *> rawbits(m_channels);
+    std::vector<double> ch(m_channels);
+    double sunz_val = 0.0;
+    mu::Parser p;
+
     for (size_t i = 0; i < m_channels; i++) {
-        rawbits[i] = (quint16 *)rawChannels[i].bits();
+        p.DefineVar("ch" + std::to_string(i+1), &ch[i]);
     }
-    std::tuple<QRgba64 *, quint16 *> bits = std::make_tuple(
-        reinterpret_cast<QRgba64 *>(image.bits()),
-        reinterpret_cast<quint16 *>(image.bits())
-    );
-
-    #pragma omp parallel
```

Comment:
Preset calculation should me multithreaded but there issues with QImage::scanLine() when calling from multiple threads